pkg/blockbuilder/blockbuilder.go

```diff
@@ -239,3 +240,308 @@ func (b *BlockBuilder) getLagForPartition(ctx context.Context, partition int32)
 
 	return kadm.GroupMemberLag{}, lastErr
 }
+
+type partitionState struct {
+	Partition          int32
+	Lag                int64
+	Commit             kadm.Offset
+	CommitRecTimestamp time.Time
+	SeenTillOffset     int64
+	LastBlockEnd       int64
+}
+
+func partitionStateFromLag(logger log.Logger, lag kadm.GroupMemberLag, fallbackMillis int64) *partitionState {
+	commitRecTs, seenTillOffset, lastBlockEnd, err := unmarshallCommitMeta(lag.Commit.Metadata)
+	if err != nil {
+		// If there is an error in unmarshalling the metadata, treat it as if
+		// we have no commit. There is no reason to stop the cycle for this.
+		level.Warn(logger).Log("msg", "error unmarshalling commit metadata", "err", err, "partition", lag.Partition, "offset", lag.Commit.At, "metadata", lag.Commit.Metadata)
+	}
+
+	if commitRecTs == 0 {
+		// If there was no commit metadata, we use the fallback to replay a set amount of
+		// records because it is non-trivial to peek at the first record in a partition to determine
+		// the range of replay required. Without knowing the range, we might end up trying to consume
+		// a lot of records in a single partition consumption call and end up in an OOM loop.
+		// TODO(codesome): add a test for this.
+		commitRecTs = fallbackMillis
+	}
+
+	return &partitionState{
+		Partition:          lag.Partition,
+		Lag:                lag.Lag,
+		Commit:             lag.Commit,
+		CommitRecTimestamp: time.UnixMilli(commitRecTs),
+		SeenTillOffset:     seenTillOffset,
+		LastBlockEnd:       lastBlockEnd,
+	}
+}
+
+func (b *BlockBuilder) consumePartition(ctx context.Context, state *partitionState, cycleEnd time.Time) error {
+	lagging := cycleEnd.Sub(state.CommitRecTimestamp) > 3*b.cfg.ConsumeInterval/2
+	if !lagging {
+		// Either we did not find a commit offset or we are not lagging behind by
+		// more than 1.5 times the consume interval.
+		// When there is no kafka commit, we play safe and assume seenTillOffset and
+		// lastBlockEnd was 0 to not discard any samples unnecessarily.
+		err := b.consumePartitionCycle(ctx, state, cycleEnd)
+		if err != nil {
+			return fmt.Errorf("consume partition %d: %w", state.Commit.Partition, err)
+		}
+
+		return nil
+	}
+
+	// We are lagging behind. We need to consume the partition in sections.
```

Comment:
nitpick on the comment - you use "cycle" more often then "section", let's stick to one term - "cycle"?

Reply:
Here a "section" means a part of a cycle (but we used to follow the kgo's vocabulary, and assumed a "part" means "partition"). That is, a "cycle" is an interval from `[T, cycleEnd)`. Sections of the cycle are chunks `[T, T+interval), [T+interval, T+2*interval), ... [T+S*interval, cycleEnd)` (assuming the duration between `cycleEnd` and `T` is N hours).

Reply:
thanks for explaining 💡